MoltenVK/MoltenVK/GPUObjects/MVKSync.h

```diff
@@ -636,3 +636,56 @@ class MVKMetalCompiler : public MVKBaseObject {
 	std::string _compilerType = "Unknown";
 	MVKPerformanceTracker* _pPerformanceTracker = nullptr;
 };
+
+#pragma mark -
+#pragma mark MVKDeferredOperation
+
+/** Defines the function pointer for each dependent function*/
+union MVKDeferredOperationFunctionPointer
+{
+    // Empty until deferred functions from other extensions have been defined
+    // Planning to use std::functions
+};
+
+/** Manages what function is being deferred*/
+enum MVKDeferredOperationFunctionType
+{
+    // Empty until deferred functions from other extensions have been defined
+};
+
+class MVKDeferredOperation : public MVKVulkanAPIDeviceObject {
+public:
+    /** Returns the Vulkan type of this object. */
+    VkObjectType getVkObjectType() override { return VK_OBJECT_TYPE_DEFERRED_OPERATION_KHR; }
+    
+    /** Returns the debug report object type of this object. */
+    VkDebugReportObjectTypeEXT getVkDebugReportObjectType() override { return VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT; }
+    
+    /** Joins the deferred operation */
+    VkResult join();
+    
+    /** Gets the max number of threads that can execute the deferred operation concurrently*/
+    uint32_t getMaxConcurrency() { return 1; } // Perhaps the number of CPU cores
+    
+    /** Gets the result of the execution of the deferred operation */
+    VkResult getResult() { return operationResult; }
+    
+    /** Sets all the variables needed for a deferred operation, however should never be called manually and only from other functions that take deferred operations*/
+    void deferOperation(MVKDeferredOperationFunctionPointer pointer, MVKDeferredOperationFunctionType type, std::vector<void*> parameters);
+#pragma mark Construction
+    MVKDeferredOperation(MVKDevice* device) : MVKVulkanAPIDeviceObject(device) {}
+protected:
+    /** Stores the result of the operation*/
+    VkResult operationResult = VK_SUCCESS;
+    
+    /** Stores a pointer to the function*/
+    MVKDeferredOperationFunctionPointer functionPointer;
+    
+    /** Stores what functions is being deferred*/
+    MVKDeferredOperationFunctionType functionType;
+    
+    /** The parameters in the operation being deferred*/
+    std::vector<void*> functionParameters = {};
```

Comment:
> I don't expect many deferred operations to have lots of parameters.

Hmmm...in that case, since this is all internalized and based on specifically adding functions to the union from later extensions, another option would be to use a basic:

```
static kMVKMaxDeferredFuncParams = 3;
void* functionParameters[kMVKMaxDeferredFuncParams];
```

And then we just update `kMVKMaxFunctionParameters` as we add additional extensions that contain deferred functions that require more parameters.

Not sure if that makes it easier, or just complicates things.

Reply:
Did you mean to use a C-Style array?

Reply:
> Did you mean to use a C-Style array?

Yes. I did. I was just reflecting on @cdavis5e observation that the number of function args will be very small, so it's not expensive to over-allocate by one or two extra pointer-size elements, when function might have less than the max number of arguments, as defined by the function with the overall most arguments.

If the largest function we support has say 4 arguments, and another function only uses only 2, we're only over-allocating by 16 bytes, which is not a lot (and is about the same as the overhead of `std::vector` or `MVKSmallVector`).

I'm just throwing the idea out as a design option, to get around the problem you're having with `MVKSmallVector<void*>`. The primary advantage to using either vector is the ability to expand at will, but we really don't need to worry about that in this situation.

All that said, the preferred MoltenVK/C++ solution would be to get `MVKSmallVector/MVKArrayRef` working!

Reply:
Thank you for the detailed answer. I've decided to use a C-Style array in the end, as the parameter count should not be that high, and the parameters will not have to resize, so a C array seems to work well for now.